subsys/mgmt/mcumgr/grp/enum_mgmt/src/enum_mgmt.c

```diff
@@ -0,0 +1,420 @@
+
+#include <mgmt/mcumgr/util/zcbor_bulk.h>
+
+#include <zcbor_common.h>
+#include <zcbor_encode.h>
+#include <zcbor_decode.h>
+
+#if defined(CONFIG_MCUMGR_MGMT_NOTIFICATION_HOOKS)
+#include <zephyr/mgmt/mcumgr/mgmt/callbacks.h>
+#include <mgmt/mcumgr/transport/smp_internal.h>
+#endif
+
+LOG_MODULE_REGISTER(mcumgr_enum_grp, CONFIG_MCUMGR_GRP_ENUM_LOG_LEVEL);
+
+struct enum_mgmt_list_arg {
+	bool *ok;
+	zcbor_state_t *zse;
+};
+
+struct enum_mgmt_details_arg {
+	bool *ok;
+	zcbor_state_t *zse;
+	uint16_t *allowed_group_ids;
+	uint16_t allowed_group_id_size;
+#if defined(CONFIG_MCUMGR_GRP_ENUM_DETAILS_HOOK)
+	enum mgmt_cb_return status;
+	int32_t err_rc;
+	uint16_t err_group;
+#endif
+};
+
+static bool enum_mgmt_cb_count(const struct mgmt_group *group, void *user_data)
+{
+	uint16_t *count = (uint16_t *)user_data;
+
+	++*count;
+
+	return true;
+}
+
+static bool enum_mgmt_cb_list(const struct mgmt_group *group, void *user_data)
+{
+	struct enum_mgmt_list_arg *data = (struct enum_mgmt_list_arg *)user_data;
+
+	*data->ok = zcbor_uint32_put(data->zse, group->mg_group_id);
+
+	return *data->ok;
+}
+
+#ifdef CONFIG_MCUMGR_GRP_ENUM_DETAILS
+static bool enum_mgmt_cb_details(const struct mgmt_group *group, void *user_data)
+{
+	struct enum_mgmt_details_arg *data = (struct enum_mgmt_details_arg *)user_data;
+
+#ifdef CONFIG_MCUMGR_GRP_ENUM_DETAILS_NAME
+	const char *group_name = group->mg_group_name;
+	int group_name_size = 0;
+#endif
+
+#if defined(CONFIG_MCUMGR_GRP_ENUM_DETAILS_HOOK)
+	struct enum_mgmt_detail_output group_detail_data = {
+		.group = group,
+		.zse = data->zse,
+	};
+#endif
+
+	if (data->allowed_group_ids != NULL && data->allowed_group_id_size > 0) {
+		/* Check if this entry is in the allow list */
+		uint16_t i = 0;
+
+		while (i < data->allowed_group_id_size) {
+			if (data->allowed_group_ids[i] == group->mg_group_id) {
+				break;
+			}
+
+			++i;
+		}
+
+		if (i == data->allowed_group_id_size) {
+			*data->ok = true;
+			goto finished;
+		}
+	}
+
+#ifdef CONFIG_MCUMGR_GRP_ENUM_DETAILS_NAME
+	if (group_name != NULL) {
+		group_name_size = strlen(group_name);
+	}
+#endif
+
+	*data->ok = zcbor_map_start_encode(data->zse, CONFIG_MCUMGR_GRP_ENUM_DETAILS_STATES) &&
+		    zcbor_tstr_put_lit(data->zse, "group") &&
+		    zcbor_uint32_put(data->zse, group->mg_group_id) &&
+#ifdef CONFIG_MCUMGR_GRP_ENUM_DETAILS_NAME
+		    zcbor_tstr_put_lit(data->zse, "name") &&
+		    zcbor_tstr_encode_ptr(data->zse, group_name, group_name_size) &&
+#endif
+#ifdef CONFIG_MCUMGR_GRP_ENUM_DETAILS_HANDLERS
+		    zcbor_tstr_put_lit(data->zse, "handlers") &&
+		    zcbor_uint32_put(data->zse, group->mg_handlers_count) &&
+#endif
+		    1;
+
+#if defined(CONFIG_MCUMGR_GRP_ENUM_DETAILS_HOOK)
+	/* Send notification to application to optionally append more fields */
+	data->status = mgmt_callback_notify(MGMT_EVT_OP_ENUM_MGMT_DETAILS, &group_detail_data,
+				      sizeof(group_detail_data), &data->err_rc, &data->err_group);
+
+	if (data->status != MGMT_CB_OK) {
```

Comment:
Not sure this will ever be used. What would be the usercase?

Reply:
To add anything that users need. An example might be an access feature, maybe they want to protect certain functions depending upon a "user level", so the minimum user level required to use it can be added

Reply:
Discussed offline, nothing against.